packages/flutter_tools/lib/src/flutter_cache.dart

```diff
@@ -846,12 +849,19 @@ class IosUsbArtifacts extends CachedArtifact {
 // TODO(zanderso): upload debug desktop artifacts to host-debug and
 // remove from existing host folder.
 // https://github.com/flutter/flutter/issues/38935
-const List<List<String>> _windowsDesktopBinaryDirs = <List<String>>[
-  <String>['windows-x64', 'windows-x64-debug/windows-x64-flutter.zip'],
-  <String>['windows-x64', 'windows-x64/flutter-cpp-client-wrapper.zip'],
-  <String>['windows-x64-profile', 'windows-x64-profile/windows-x64-flutter.zip'],
-  <String>['windows-x64-release', 'windows-x64-release/windows-x64-flutter.zip'],
-];
+
+List<List<String>> _getWindowsDesktopBinaryDirs() {
+  return <List<String>>[
+    <String>['windows-x64', 'windows-x64-debug/windows-x64-flutter.zip'],
+    <String>['windows-x64', 'windows-x64/flutter-cpp-client-wrapper.zip'],
+    <String>['windows-x64-profile', 'windows-x64-profile/windows-x64-flutter.zip'],
+    <String>['windows-x64-release', 'windows-x64-release/windows-x64-flutter.zip'],
+    <String>['windows-arm64', 'windows-arm64-debug/windows-arm64-flutter.zip'],
```

Comment:
This part is more debatable, as it means we download arm64 and x64 flutter engine for any build. Technically, we could download only "target" architecture, but I didn't started this work, because pushing information until this part of the code is really intrusive (lots of layers to cross).

Any idea how we could set this somewhere?

Reply:
I explicitely opened this conversation to ask for guidance on how to implement this, not to ask "How hard would it be?". The problem is that code is structured this way:
windows_stuff -> generic code -> windows_download_specific

From what I understood, there is no details about `target-platform` in the forwarded structure. If I had this, it will be a nightmare to update generic part, without thinking about all the tests to update. So that's why I'm asking how to implement this cleanly (i.e. not with a global variable) with a minimal impact on all layers crossed.

Reply:
> From what I understood, there is no details about target-platform in the forwarded structure. 
This is correct.

> Hm... How hard would it be to download Windows Arm64 artifacts only if we're currently on the master release channel? If that's easier, I'd be okay with punting this problem down the road.

I agree with @loic-sharma here. A nuance though is, how do we detect from tool code whether or not Windows arm64 is available? I think the way we would usually do this is to add a [feature flag](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/features.dart#L64) that would be *available* on whatever channels we want this available to (is it currently only "master"?). It's ugly, but I would be satisfied to insert a call to `globals.features.isWindowsArmEnabled` here and conditionally add the artifacts here.

Reply:
Based on current `dart.exe`, we can detect if Abi is windows-x64 or windows-arm64.
This binary is downloaded by `update_dart_sdk.ps1`, which will check PROCESSOR_ARCHITECTURE env var.

For the question you ask, more precisely, there is no way to detect if engine is available or not, given a specific revision. Either we introduce a mechanism that "tries to" download, and fallback to x64 if it fails (eyes already rolling here), or we simply know that master channel supports it, because Dart team guarantees this. It would be much easier if all channels would support it, but alas, it can't be done, because Dart team has no access to `windows-arm64` hardware, from what I understood.

We can Introduce a feature flag, but I don't really see which value it will bring. You can have it `true`, and still only want windows-x64 artifacts. The problem is mostly architectural, that Flutter tool never had to do this kind of selective download until now.
`flutter build linux` avoided the problem, by simply [deactivating this kind of cross compilation](https://github.com/flutter/flutter/blob/a9ae39df677bd3b1fe01b5c83a321e4baa6fe6bc/packages/flutter_tools/lib/src/commands/build_linux.dart#L76)...

---

Even if we had a global variable TargetPlatform? in globals.dart, it would not solve everything, because flutter sdk is downloaded *BEFORE* any argument is parsed, so we don't know at this point if windows-arm64 and windows-x64 are both needed or not.

That's why I'm asking how to update flutter cache to match some given architecture, without having to modify 5 layers of code, and change all unit tests associated.

Reply:
Does checking the current channel, and download both architecture only on `master` channel is ok for you?

Reply:
Updated this PR to only do native compilation. Thus, we only download native artifacts, and there is no need to have both architectures, nor to check channel.